### PR TITLE
Kubernetes 1.2.2 update

### PR DESCRIPTION
**Pull request by mbruzek**
For your consideration I bring you the kubernetes 1.2.2 update!

The template update by @eddytruyen opened a can-of-worms! I could not validate the PR template update with skydns because the services was not coming up correctly.

At one point I decided to update all the templates to the latest release of Kubernetes (which in hindsight was a mistake) based off of the skydns template stuff.

This code is working I was able to get a successful nslookup from a busybox container as described here: https://github.com/kubernetes/kubernetes/tree/release-1.2/cluster/addons/dns#how-do-i-test-if-it-is-working

```
root@ip-172-31-17-171:/home/ubuntu# kubectl exec busybox -- nslookup kubernetes.default
Server:    10.1.0.10
Address 1: 10.1.0.10 ip-10-1-0-10.ec2.internal

Name:      kubernetes.default
Address 1: 10.1.0.1 ip-10-1-0-1.ec2.internal
root@ip-172-31-17-171:/home/ubuntu# kubectl exec busybox -it -- /bin/sh
/ # 
/ # nslookup kubernetes.default
Server:    10.1.0.10
Address 1: 10.1.0.10 ip-10-1-0-10.ec2.internal

Name:      kubernetes.default
Address 1: 10.1.0.1 ip-10-1-0-1.ec2.internal
/ # nslookup kubernetes
Server:    10.1.0.10
Address 1: 10.1.0.10 ip-10-1-0-10.ec2.internal

Name:      kubernetes
Address 1: 10.1.0.1 ip-10-1-0-1.ec2.internal
/ # exit
```

Please have a look, I know this seems like a lot of code changed, but the update was necessary. Please also build and deploy kubernetes and relate to etcd charm to test this yourself using the instructions above.

### CODE REVIEW

**Comment by eddytruyen** [comment]
This is great! How can I deploy these changes in my juju environment?

I am used to deploy the kubernetes-core bundle using quickstart: `juju quickstart u/containers/kubernetes-core/0`, but it seems this bundle does not reflect the latest changes.

Thanks for the great effort!

**Comment by mbruzek** [comment]
@eddytruyen you have to pull my github branch and `charm build` the charm with charm tools 2.1.2. Then change the bundle to use a local:trusty/kubernetes charm rather than cs:~kubernetes/trusty/kubernetes (cs stands for the Charm Store). Alternately you can do the following:

Start with the Juju docs on [Creating a new Layer](https://jujucharms.com/docs/devel/developer-getting-started#creating-a-new-layer) to ensure you have the right environment variables set.

```
git clone https://github.com/mbruzek/layer-k8s.git
cd layer-k8s
git checkout kubernetes-1.2.2-update
charm build -l DEBUG
juju deploy cs:trusty/etcd
juju deploy local:trusty/kubernetes
juju add-relation etcd kubernetes
```

Wait for that process to complete, and you should have v1.2.3 of kubernetes with skydns running with a single etcd unit. Please let me know if you have any problems or find any bugs.

Thanks.

File: README.md [hunk 1]
```diff
@@ -65,7 +68,7 @@ cluster is operational.
 
  - [Kubernetes github project](https://github.com/kubernetes/kubernetes)
  - [Kubernetes issue tracker](https://github.com/kubernetes/kubernetes/issues)
- - [Kubernetes Documenation](https://github.com/kubernetes/kubernetes/tree/master/docs)
```
Comment (lazypower) [inline README.md:68]: K8s :cop: approves of this change

File: reactive/k8s.py [hunk 2]
```diff
@@ -283,39 +281,44 @@ def render_files(reldata=None):
...
-    store = unitdata.kv()
-    certificate_data = store.get('tls.{0}.certificate'.format(prefix))
-    certificate_name = '{0}.crt'.format(prefix)
-    # The certificate should be saved to this directory.
-    certificate_path = os.path.join(directory, certificate_name)
-    # write the server certificate out to the correct location
-    with open(certificate_path, 'w') as fp:
-        fp.write(certificate_data)
+def arch():
+    '''Return the package archiecture as a string. Raise an exception if the
```
Comment (lazypower) [inline reactive/k8s.py]: s/archiecture/architecture/

**Comment by lazypower** [comment]
This PR LGTM, I still need to build / deploy, but prelim glance is looking great. only minor nits/comments left inline

**Comment by lazypower** [comment]
One additional nit - with the upcoming 2.0 changes local charms are basically no longer a thing.. we'll need to get series declared in metadata

ubuntu@devbox:~$ juju deploy ./trusty/kubernetes
ERROR series not specified and charm does not define any

**Comment by mbruzek** [comment]
I was not aware that juju 1.x supported series yet, but you told me it does.

**Comment by lazypower** [comment]
+1 LGTM

:shipit:

**Comment by eddytruyen** [comment]
[juju-logs-on-machine8-kubernetes-unit.zip](https://github.com/mbruzek/layer-k8s/files/241649/juju-logs-on-machine8-kubernetes-unit.zip)
@mbruzek 
For some reason; the juju agent dies on every machine to which I deploy a kubernetes-unit.
I am using the latest agent-version on an ubuntu 14.04 machine.

I get the following debug-output:

```
unit-kubernetes-1[7864]: 2016-04-28 21:41:32 WARNING juju.worker.dependency engine.go:304 failed to start "uniter" manifold worker: dependency not available
unit-kubernetes-1[7864]: 2016-04-28 21:41:35 ERROR juju.worker.uniter.filter filter.go:137 tomb: dying
```

I attached the logs for /var/logs/juju/. Can you please have a look at this because 9 out of 10 add-unit commands fail in the same way,

**Comment by eddytruyen** [comment]
@mbruzek I continued retrying adding kubernetes-units until one succeeded. (But still 9 out of 10 add-unit commands fail as described in my previous comment).

I did the dns test but it was not completely succesfull.

```
/ # nslookup kubernetes.default
Server:    10.1.5.10
Address 1: 10.1.5.10

nslookup: can't resolve 'kubernetes.default'
```

A single nslookup takes aproximately a minute.

logs from healthz:
`
2016/04/29 00:00:22 Client ip 10.1.5.1:44498 requesting /healthz probe servici
ng cmd nslookup kubernetes.default.svc.cluster.local 127.0.0.1 >/dev/null
`
logs from kube2sky:

```
ubuntu@juju-openstack-machine-1:~$ kubectl logs kube-dns-v11-opix2 -c kube2sky --namespace=kube-system | more
W0428 23:54:33.711133       1 flags.go:39] kube_master_url is DEPRECATED and will be remo
ved in a future version. Use kube-master-url instead.
I0428 23:54:33.717403       1 kube2sky.go:462] Etcd server found: http://127.0.0.1:4001
I0428 23:54:34.721715       1 kube2sky.go:529] Using http://172.17.13.194:8080 for kubern
etes master
I0428 23:54:34.721904       1 kube2sky.go:530] Using kubernetes API v1
I0428 23:54:34.722224       1 kube2sky.go:598] Waiting for service: default/kubernetes
I0428 23:54:35.297341       1 kube2sky.go:660] Successfully added DNS record for Kubernet
es service.
```

Logs from skydns:

```
2016/04/28 23:54:33 skydns: falling back to default configuration, could not read from et
cd: 100: Key not found (/skydns/config) [24]
2016/04/28 23:54:33 skydns: ready for queries on cluster.local. for tcp://0.0.0.0:53 [rca
che 0]
2016/04/28 23:54:33 skydns: ready for queries on cluster.local. for udp://0.0.0.0:53 [rca
che 0]
```

Logs from etcd

```
ubuntu@juju-openstack-machine-1:~$ kubectl logs kube-dns-v11-opix2 -c etcd --namespace=kube-system | more
2016-04-28 23:54:33.426227 I | etcdmain: etcd Version: 2.2.1
2016-04-28 23:54:33.426303 I | etcdmain: Git SHA: 75f8282
2016-04-28 23:54:33.426308 I | etcdmain: Go Version: go1.5.1
2016-04-28 23:54:33.426312 I | etcdmain: Go OS/Arch: linux/amd64
2016-04-28 23:54:33.438424 I | etcdmain: setting maximum number of CPUs to 1, total numbe
r of available CPUs is 1
2016-04-28 23:54:33.439267 N | etcdmain: the server is already initialized as member befo
re, starting as etcd member...
2016-04-28 23:54:33.439577 I | etcdmain: listening for peers on http://localhost:2380
2016-04-28 23:54:33.439643 I | etcdmain: listening for peers on http://localhost:7001
2016-04-28 23:54:33.439676 I | etcdmain: listening for client requests on http://127.0.0.
1:2379
2016-04-28 23:54:33.439706 I | etcdmain: listening for client requests on http://127.0.0.
1:4001
2016-04-28 23:54:33.446836 I | etcdserver: name = default
2016-04-28 23:54:33.446902 I | etcdserver: data dir = /var/etcd/data
2016-04-28 23:54:33.446954 I | etcdserver: member dir = /var/etcd/data/member
2016-04-28 23:54:33.446986 I | etcdserver: heartbeat = 100ms
2016-04-28 23:54:33.447030 I | etcdserver: election = 1000ms
2016-04-28 23:54:33.447061 I | etcdserver: snapshot count = 10000
2016-04-28 23:54:33.447113 I | etcdserver: advertise client URLs = http://127.0.0.1:2379,
http://127.0.0.1:4001
2016-04-28 23:54:33.488518 I | etcdserver: restarting member 6a5871dbdd12c17c in cluster
f68652439e3f8f2a at commit index 4548
2016-04-28 23:54:33.488910 I | raft: 6a5871dbdd12c17c became follower at term 2
2016-04-28 23:54:33.488997 I | raft: newRaft 6a5871dbdd12c17c [peers: [], term: 2, commit
: 4548, applied: 0, lastindex: 4548, lastterm: 2]
2016-04-28 23:54:33.518192 I | etcdserver: starting server... [version: 2.2.1, cluster ve
rsion: to_be_decided]
2016-04-28 23:54:33.538174 N | etcdserver: added local member 6a5871dbdd12c17c [http://lo
calhost:2380 http://localhost:7001] to cluster f68652439e3f8f2a
2016-04-28 23:54:33.538369 N | etcdserver: set the initial cluster version to 2.2
2016-04-28 23:54:35.218284 I | raft: 6a5871dbdd12c17c is starting a new election at term
2
2016-04-28 23:54:35.218491 I | raft: 6a5871dbdd12c17c became candidate at term 3
2016-04-28 23:54:35.218513 I | raft: 6a5871dbdd12c17c received vote from 6a5871dbdd12c17c
 at term 3
2016-04-28 23:54:35.218556 I | raft: 6a5871dbdd12c17c became leader at term 3
2016-04-28 23:54:35.218607 I | raft: raft.node: 6a5871dbdd12c17c elected leader 6a5871dbd
d12c17c at term 3
2016-04-28 23:54:35.287686 I | etcdserver: published {Name:default ClientURLs:[http://127
.0.0.1:2379 http://127.0.0.1:4001]} to cluster f68652439e3f8f2a
```

**Comment by mbruzek** [comment]
@eddytruyen regarding the juju agent issue. I don't know how that would be related to kubernetes. The log you attached did not contain any information about the problem. Can you reproduce this problem with any other (more simple) setups? If you still see the error, please file a bug with the juju-core project: https://bugs.launchpad.net/juju-core/+filebug They will likely need the machine logs along with the unit logs.

regarding the nslookp failure. I tested this on juju 1.25.5 and 2.0.0-beta5 both worked with the busybox/nslookup on AWS. I see you are using an openstack provider, and perhaps there is something different with that provider that is causing this problem. I don't have an openstack environment to test with at this time (but can look into it). Again I don't see anything obvious in the sections of logs you included, so more information would be helpful.

**Pull request merged**